### PR TITLE
Enable search scale for awq

### PR DESCRIPTION
**Pull request by AllentDan**
- [x] Add search scale option for awq
- [x] Add batch size flag

### CODE REVIEW

File: lmdeploy/lite/apis/calibrate.py [hunk 1]
```diff
@@ -118,7 +119,11 @@ def calibrate(model: str,
               calib_samples: int = 128,
               calib_seqlen: int = 2048,
               work_dir: str = './work_dir',
-              device: str = 'cuda') -> None:
+              device: str = 'cuda',
+              w_bits: int = 4,
+              w_group_size: int = 128,
+              search_scale: bool = False,
+              batch_size: bool = 1) -> None:
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:126]: typo: bool -> int
Is `1` a good default value?

File: lmdeploy/lite/apis/calibrate.py [hunk 2]
```diff
@@ -118,7 +119,11 @@ def calibrate(model: str,
               calib_samples: int = 128,
               calib_seqlen: int = 2048,
               work_dir: str = './work_dir',
-              device: str = 'cuda') -> None:
+              device: str = 'cuda',
+              w_bits: int = 4,
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:123]: I don't think we should expose `w_bits` since we only support 4bit weight-only quantization now.

File: lmdeploy/lite/apis/calibrate.py [hunk 3]
```diff
@@ -134,6 +139,11 @@ def calibrate(model: str,
             Defaults to './work_dir'.
         device (str, optional): The device to be used for calculation.
             Defaults to 'cuda'.
+        w_bits (int): Bit number for weight quantization.
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:142]: miss `search_scale`

File: lmdeploy/lite/apis/calibrate.py [hunk 4]
```diff
@@ -118,7 +119,11 @@ def calibrate(model: str,
               calib_samples: int = 128,
               calib_seqlen: int = 2048,
               work_dir: str = './work_dir',
-              device: str = 'cuda') -> None:
+              device: str = 'cuda',
+              w_bits: int = 4,
+              w_group_size: int = 128,
+              search_scale: bool = False,
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:125]: In what cases we are supposed to suggest users enabling `search_scale`?
Should we set the default value True?

File: lmdeploy/lite/quantization/awq.py [hunk 5]
```diff
@@ -251,3 +251,62 @@ def smooth_layers(layers,
...
+    if w_group_size > 0:
+        assert org_w_shape[-1] % w_group_size == 0
+        w = w.reshape(-1, w_group_size)
+    assert w.dim() == 2
+    max_val = w.amax(dim=1, keepdim=True)
+    min_val = w.amin(dim=1, keepdim=True)
+    max_int = 2**w_bit - 1
+    min_int = 0
+    scales = (max_val - min_val).clamp(min=1e-5) / max_int
+    zeros = (-torch.round(min_val / scales)).clamp_(min_int, max_int)
```
Comment (pppppM) [inline lmdeploy/lite/quantization/awq.py:268]: 直接使用 torch round 对数值会有一些影响
torch.round(1.5) = 2
torch.round(2.5) = 2

可以使用 precise_round 接口
https://github.com/InternLM/lmdeploy/blob/a4344291ded68e552a5070f552b2e482ea7b43ff/lmdeploy/lite/utils/cal_qparams.py#L14-L16

File: lmdeploy/lite/quantization/awq.py [hunk 6]
```diff
@@ -251,3 +251,62 @@ def smooth_layers(layers,
...
+    assert torch.isnan(scales).sum() == 0
+    assert torch.isnan(w).sum() == 0
+
+    w = (torch.clamp(torch.round(w / scales) + zeros, min_int, max_int) -
```
Comment (pppppM) [inline lmdeploy/lite/quantization/awq.py:272]: torch.round 数值问题

File: lmdeploy/lite/apis/calibrate.py [hunk 7]
```diff
@@ -182,11 +192,23 @@ def calibrate(model: str,
                                         seqlen=calib_seqlen)
 
     # Initialize calibration context
-    calib_ctx = CalibrationContext(model,
-                                   tokenizer,
-                                   layer_type=layer_type,
-                                   norm_type=norm_type,
-                                   device=device)
+    if search_scale:
+        calib_ctx = AWQCalibrationContext(model,
```
Comment (pppppM) [inline lmdeploy/lite/apis/calibrate.py:196]: 现在的 search scale  w8a8 smooth quant 可以用么？如果可以用的话可以改个更通用一点的名字

**Comment by lvhan028** [comment]
有尝试过开启search_scale，量化 codellama 么？

**Comment by lvhan028** [comment]
https://github.com/InternLM/lmdeploy/issues/469

File: lmdeploy/lite/quantization/awq.py (hunk 5 above)
Comment (lzhangzz) [reply lmdeploy/lite/quantization/awq.py:268]: `torch.round` 符合 IEEE 754 默认的浮点舍入操作

https://en.wikipedia.org/wiki/Rounding#Rounding_half_to_even

`x.sign() * (x.abs() + 0.5).floor()` 作用在向量上貌似会使 norm 的期望变大

File: lmdeploy/lite/apis/calibrate.py (hunk 1 above)
Comment (AllentDan) [reply lmdeploy/lite/apis/calibrate.py:126]: 1 is for BC breaking, the previous design is for low memory users.

File: lmdeploy/lite/apis/calibrate.py (hunk 4 above)
Comment (AllentDan) [reply lmdeploy/lite/apis/calibrate.py:125]: When smooth quant model accuracy is not satisfying, we recommend it. Or, when a user does not care about time consuming of `search_scale`.

File: lmdeploy/lite/quantization/awq.py (hunk 5 above)
Comment (pppppM) [reply lmdeploy/lite/quantization/awq.py:268]: 如果计算 scale 或 zeros 时用 torch.round，会触发一些 clamp，导致量化后的权重，与按照公式手动算出来的结果不一样，会有 > 15 的数

如果是按照量化公式手算的话，不需要 clamp，最终的 w 也不会有 > 15 的数；
早期在 qwen1 上遇到过这种问题 #605 ，通过 precise_round 来保证和手算结果一致，解决了 qwen1 量化掉点的问题

File: lmdeploy/lite/apis/calibrate.py (hunk 7 above)
Comment (AllentDan) [reply lmdeploy/lite/apis/calibrate.py:198]: 可以用，但是不知道什麼名字合適

**Comment by AllentDan** [comment]
> 有尝试过开启search_scale，量化 codellama 么？#469

試了下 codellama-instruct-7b，稍微好一些，不顯著
| dataset | version | metric | mode    | codellama-7b-chat | codellama-7b-chat-4bit | codellama-7b-chat-4bit-noscale |
|---------|---------|--------|--------|-------------------|----------------------|------------------------------|
| ceval   | -       | naive_average | gen   | 34.35             | 32.71                 | 34.61                        |
| mmlu    | -       | naive_average | gen   | 35.28             | 33.85                 | 31.99                        |
| triviaqa | 2121ce  | score          | gen   | 39.58             | 36.63                 | 35.42                        |

**Comment by lvhan028** [comment]
我记得在之前的版本中，量化codellama会直接报错的，好像是有NAN

File: lmdeploy/lite/apis/calibrate.py (hunk 7 above)
Comment (lvhan028) [reply lmdeploy/lite/apis/calibrate.py:198]: how about CalibrationContextV2?

**Comment by lvhan028** [comment]
@zhulinJulia24 pls launch the evaluation test on llama2-7b, llama3-8b, internlm2-7b, internlm2-20b, qwen-7b, qwen1.5-7b, codellama-7b

**Review by lvhan028 — APPROVED** [review]

**Comment by zhulinJulia24** [comment]
lmdeploy lite auto_awq /nvme/qa_test_models/Qwen/Qwen-7B-Chat --work-dir /nvme/qa_test_models/Qwen/Qwen-7B-Chat-inner-4bits --batch-size 3 --search-scale True
Try importing flash-attention for faster inference...
Warning: import flash_attn rms_norm fail, please install FlashAttention layer_norm to get higher efficiency [flash-attention/csrc/layer_norm at main · Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention/tree/main/csrc/layer_norm)
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:02<00:00,  2.70it/s]
Move transformer.wte to GPU.
Move transformer.drop to GPU.
Move transformer.rotary_emb to GPU.
Move transformer.h.0 to CPU.
Move transformer.h.1 to CPU.
Move transformer.h.2 to CPU.
Move transformer.h.3 to CPU.
Move transformer.h.4 to CPU.
Move transformer.h.5 to CPU.
Move transformer.h.6 to CPU.
Move transformer.h.7 to CPU.
Move transformer.h.8 to CPU.
Move transformer.h.9 to CPU.
Move transformer.h.10 to CPU.
Move transformer.h.11 to CPU.
Move transformer.h.12 to CPU.
Move transformer.h.13 to CPU.
Move transformer.h.14 to CPU.
Move transformer.h.15 to CPU.
Move transformer.h.16 to CPU.
Move transformer.h.17 to CPU.
Move transformer.h.18 to CPU.
Move transformer.h.19 to CPU.
Move transformer.h.20 to CPU.
Move transformer.h.21 to CPU.
Move transformer.h.22 to CPU.
Move transformer.h.23 to CPU.
Move transformer.h.24 to CPU.
Move transformer.h.25 to CPU.
Move transformer.h.26 to CPU.
Move transformer.h.27 to CPU.
Move transformer.h.28 to CPU.
Move transformer.h.29 to CPU.
Move transformer.h.30 to CPU.
Move transformer.h.31 to CPU.
Move transformer.ln_f to GPU.
Move lm_head to CPU.
Loading calibrate dataset ...
**Token indices sequence length is longer than the specified maximum sequence length for this model (1104485 > 32768). Running this sequence through the model will result in indexing errors**


qwen量化过程会有这个报错，如果没有 search-scale和batch size是没有报错的。 @AllentDan

**Comment by zhulinJulia24** [comment]
测试了llama3，internlm2和qwen1.5 提升不是很显著，https://github.com/zhulinJulia24/lmdeploy/actions/runs/9015990252

原来的 https://github.com/zhulinJulia24/lmdeploy/actions/runs/9042418710
image
image



这个pr的 https://github.com/InternLM/lmdeploy/actions/runs/9061410958
image

**Comment by zhulinJulia24** [comment]
更多数据集的结果
https://github.com/InternLM/lmdeploy/actions/runs/9065329539

**Comment by AllentDan** [comment]
@zhulinJulia24 Hi, I updated some codes. And this may help both `search_scale` and  `not search_scale`. Please give a try on this.

As for the qwen warning, it appears whether or not `search_scale` is True.

**Comment by AllentDan** [comment]
This is the performance of the new codes.
```
ceval                                   -          naive_average  gen     49.08
mmlu                                    -          naive_average  gen     64.83
WiC                                     d06864     accuracy       gen     50.94
WSC                                     7902a7     accuracy       gen     36.54
triviaqa                                2121ce     score          gen     59.85
gsm8k                                   1d7fe4     accuracy       gen     76.19
race-middle                             9a54b6     accuracy       gen     85.03
race-high                               9a54b6     accuracy       gen     79.53
```

**Comment by zhulinJulia24** [comment]
I cannot do awq on codellama? Can you have a try?

`lmdeploy lite auto_awq /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf --work-dir /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf-inner-4bits --batch-size 3 --search-scale True`

**Comment by AllentDan** [comment]
> I cannot do awq on codellama? Can you have a try? @AllentDan
> 
> `lmdeploy lite auto_awq /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf --work-dir /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf-inner-4bits --batch-size 3 --search-scale True`

What is the error log?

File: lmdeploy/legacy/pytorch/modules/linear.py [hunk 8]
```diff
@@ -106,9 +106,11 @@ def from_linear(cls: Type['WeightOnlyQLinear'],
                       group_size)
         qlinear.bias = linear.bias
 
-        if qparams is not None:
+        if qparams is None:
```
Comment (lvhan028) [inline lmdeploy/legacy/pytorch/modules/linear.py:109]: legacy 中的代码和lite模块有关？

Comment (AllentDan) [reply lmdeploy/legacy/pytorch/modules/linear.py:109]: 嗯嗯，最后一步量化用到了

Comment (lvhan028) [reply lmdeploy/legacy/pytorch/modules/linear.py:109]: 那我们需要把它提到 lite 模型中，因为legacy之后会被删除。
迁移linear.py的事情放在TODO中吧。
另外，legacy中还有其他模块被用到么？

Comment (AllentDan) [reply lmdeploy/legacy/pytorch/modules/linear.py:109]: 量化代码只有用到 legacy 中的 `WeightOnlyQLinear`

**Comment by zhulinJulia24** [comment]
> > I cannot do awq on codellama? Can you have a try? @AllentDan
> > `lmdeploy lite auto_awq /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf --work-dir /nvme/qa_test_models/codellama/CodeLlama-7b-Instruct-hf-inner-4bits --batch-size 3 --search-scale True`
> 
> What is the error log?

It seems related to my model file. After I update it, it works fine now!

File: docs/zh_cn/quantization/w4a16.md [hunk 9]
```diff
@@ -52,6 +52,10 @@ lmdeploy lite auto_awq internlm/ianternlm2-chat-7b --work-dir internlm2-chat-7b-
 我们建议 --work-dir 参数带有模型名字，就像上面的例子展示的那样。这样在推理时，就不用指定对话模板了。因为推理接口会以模糊搜索方式，选出和 --work-dir 近似的对话模板
 ```
 
+```{note}
+--search-scale 相对比较耗时，尤其是 --batch-size 是 1, 而 calib_samples 和 calib_seqlen 比较大的时候。所以我们建议用户在不关心量化模型耗时或者量化后模型精度不好的时候使用。
```
Comment (lvhan028) [inline docs/zh_cn/quantization/w4a16.md:56]: 这句note建议修改为：
```
如果量化模型精度有损，建议开启 --search-scale 重新量化，并调大 --batch-size，比如 8。
search_scale 开启后，量化过程会比较耗时。--batch-size 会影响内存占用量，可以根据实际情况，酌情调整。
```

**Comment by zhulinJulia24** [comment]
image

File: docs/en/quantization/w4a16.md [hunk 10]
```diff
@@ -56,6 +58,10 @@ lmdeploy lite auto_awq internlm/internlm2-chat-7b --work-dir internlm2-chat-7b-4
 We recommend that you specify the --work-dir parameter, including the model name as demonstrated in the example above. This facilitates LMDeploy in fuzzy matching the --work-dir with an appropriate built-in chat template. Otherwise, you will have to designate the chat template during inference.
 ```
 
+```{note}
+The --search-scale option is relatively time-consuming, especially when --batch-size is set to 1 and both calib_samples and calib_seqlen are large. Therefore, we recommend users to employ this option when they are not concerned about the time it takes to quantize the model or when the precision of the quantized model is not a priority.
```
Comment (lvhan028) [inline docs/en/quantization/w4a16.md:62]: 这句话也根据中文修改下吧

File: docs/en/quantization/w4a16.md [hunk 11]
```diff
@@ -43,6 +43,8 @@ lmdeploy lite auto_awq \
   --calib-seqlen 2048 \
   --w-bits 4 \
   --w-group-size 128 \
+  --batch-size 1 \
```
Comment (lvhan028) [inline docs/en/quantization/w4a16.md:46]: 中文文档中的命令，缺少了这两个参数
